### DISCUSSION TITLE
pipx upgrade from wheel file

### DISCUSSION THREAD

**Discussion by greywidget** [pypa/pipx#1345] [Q&A]
I have a few personal utilities that are a bit too small and specific to push to [pypi](https://pypi.org)

I install these with `pipx` from local wheel files and it works very well.

As far as I can tell, it is not possible to `pipx upgrade` from a wheel file.
It isn't a big deal to `pipx uninstall` and then reinstall again from a later wheel file, but I wondered if you had ever considered allowing `pipx upgrade` from a wheel file?

I find `pipx` to be an awesome package and very much appreciate your hard work.

Thanks kindly

**Comment by browseman** [comment]
I manage to perform the `pipx upgrade` using the following approach:
```
# If the install was performed with this command/version:
pipx install --system-site-packages ./chirp-20230509-py3-none-any.whl

# Then download the latest and rename the file:
wget https://archive.chirpmyradio.com/chirp_next/next-20240427/chirp-20240427-py3-none-any.whl
mv chirp-20240427-py3-none-any.whl chirp-20230509-py3-none-any.whl

# Perform the upgrade
pipx upgrade ./chirp-20230509-py3-none-any.whl
```

I guess a simple update of the documentation could help a lot in this direction, in the case the above process is actually valid.

**Reply by greywidget** [reply]
@browseman Thanks for your reply.

Sorry but I'm a little confused by it.
In your example above `chirp` is your package and `chirp-20230509-py3-none-any.whl` is your (renamed) wheel file. I thought `pipx upgrade` expected a package name parameter.

If I try the same thing with my package:
```
pipx upgrade pycpy-0.0.3-py3-none-any.whl
Package is not installed. Expected to find C:\Users\Craig\.local\pipx\venvs\pycpy-0-0-3-py3-none-any-whl, but it does
not exist.
```

It seems to be trying to find a `venv` named after my wheel `pycpy-0-0-3-py3-none-any-whl` rather than my package `pycpy`

I notice that you specify `--system-site-packages` on your install, which I didn't. I don't know if that is why it works for you.

To be honest, even if this solution does work, I feel like renaming the latest wheel file to an older wheel file is a bit confusing (and probably more work) than uninstalling/reinstalling.

Thanks for taking the time to respond.

**Reply by browseman** [reply]
> Sorry but I'm a little confused by it.

@greywidget Your confusion is right. I did miss a vital step in my message, which I though initially that is not needed because it did return an error. But at the end it did actually updated the spec file inside the venv (I guess) and made the next command succeed. So my actual sequence of executed commands was:
```
...

# Perform the upgrade
pipx upgrade --install ./chirp-20230509-py3-none-any.whl #this one gave me error so I did omit it in the first message.
pipx upgrade chirp
```

This it did work, but in today attempt to reproduce the behavior in a clean environment I discover that the following one works better and is not even needed to rename files:
```
pipx install --force --system-site-packages <NEW_WHL_FILE>.whl
```
> It seems to be trying to find a venv named after my wheel pycpy-0-0-3-py3-none-any-whl rather than my package pycpy

Yes, I noticed exactly the same and that's why I went in the direction to try with `pipx upgrade --install`.

> I notice that you specify --system-site-packages on your install, which I didn't. I don't know if that is why it works for you.

About the `--system-site-packages` -  it is an installation requirement (from the docs) of the application that I used `Chirp`. So most probably you don't need that for your case, at least if there isn't a specific case for it mentioned in the documentation. 

>To be honest, even if this solution does work, I feel like renaming the latest wheel file to an older wheel file is a bit confusing (and probably more work) than uninstalling/reinstalling.

I agree, it's ugly and I didn't like it as well, but it did work so that's why I shared. 
Try with the new approach that I mentioned above and write back if did worked for you. Here is the output of my attempt:
```
root@chirpupgradetest:/# chirp --version
CHIRP next-20240419 on Linux - Debian GNU/Linux trixie/sid (Python 3.11.9)

root@chirpupgradetest:/# pipx install --force --system-site-packages chirp-20240427-py3-none-any.whl
--python is ignored when --force is passed. If you want to reinstall chirp with /usr/bin/python3, run `pipx reinstall chirp-20240427-py3-none-any.whl --python /usr/bin/python3` instead.
Installing to existing venv 'chirp'
  installed package chirp 20240427, installed using Python 3.11.9
  These apps are now globally available
    - chirp
    - chirpc
done! ✨ 🌟 ✨

root@chirpupgradetest:/# chirp --version
CHIRP next-20240427 on Linux - Debian GNU/Linux trixie/sid (Python 3.11.9)

```

**Reply by greywidget** [reply]
@browseman  Thanks for your clarification, much appreciated.

I agree and can confirm that `pipx install --force <wheel file>` worked for me too:

```
pycpy --version
PyCpy 0.0.3

pipx install --force C:\Users\Craig\python_wheels\pycpy-0.0.4-py3-none-any.whl
--python is ignored when --force is passed. If you want to reinstall pycpy with
C:\Users\Craig\AppData\Local\Programs\Python\Python312\python.exe, run `pipx reinstall
C:\Users\Craig\python_wheels\pycpy-0.0.4-py3-none-any.whl --python
C:\Users\Craig\AppData\Local\Programs\Python\Python312\python.exe` instead.
Installing to existing venv 'pycpy'
⚠️  Overwriting file C:\Users\Craig\.local\bin\pycpy.exe with C:\Users\Craig\.local\pipx\venvs\pycpy\Scripts\pycpy.exe
  installed package pycpy 0.0.4, installed using Python 3.12.1
  These apps are now globally available
    - pycpy.exe
done! ✨ 🌟 ✨

pycpy --version
PyCpy 0.0.4
```

I was perusing the --help for `pipx install` and wondered if passing `--pip-args="-U"` might do it, but that doesn't work without using --force (which works anyway)

I also had a quick play with `pipx reinstall` as per the warning above, but I think that message is not quite right - I think `reinstall` expects a package, so the wheel file won't work.

So `pipx install --force` is good for me.

Thanks again for your time and information 👍

**Comment by SirHegel** [comment]
`pipx upgrade` already works from a local install — the catch is *what path* you installed from.

pipx stores the original spec and re-runs pip against it. With a wheel, that spec has the version baked into the filename:

```console
$ pipx install ./mitool-1.0.0-py3-none-any.whl
$ pipx list --json | jq -r '.venvs.mitool.metadata.main_package.package_or_url'
/tmp/wh/mitool-1.0.0-py3-none-any.whl
```

So `pipx upgrade` re-installs that exact file, no matter what else is sitting next to it:

```console
$ pipx upgrade mitool           # mitool-1.1.0-py3-none-any.whl exists in the same directory
upgrading mitool from spec '/tmp/wh/mitool-1.0.0-py3-none-any.whl'...
mitool is already at latest version 1.0.0
```

Nothing is broken — the spec is just pinned by construction.

**Point it at the source directory instead** and upgrades work the way you want:

```console
$ pipx install ./src            # stores package_or_url: /tmp/wh/src
$ # ... bump the version in pyproject.toml, edit the code ...
$ pipx upgrade mitool
upgrading mitool from spec '/tmp/wh/src'...
upgraded package mitool from 1.2.0 to 1.3.0
```

The stored spec has no version in it, so each `pipx upgrade` rebuilds from whatever is there now. For personal utilities you keep in a local checkout, that is usually what you want, and it drops the build-a-wheel step entirely.

If you would rather keep shipping wheels, `pipx install --force ./mitool-1.1.0-py3-none-any.whl` does it in one command — no `uninstall` first:

```console
$ pipx install --force ./mitool-1.1.0-py3-none-any.whl
$ mitool
mitool 1.1.0
```

A third option, if you want `pipx upgrade` *and* wheels: publish to a local directory index and install with `--pip-args="--find-links=/path/to/wheels"`, so the spec is the bare name and pip resolves the newest wheel there.

<sub>All four transcripts above are from an actual run against a throwaway `PIPX_HOME`, not from reading the code. Claude helped me trace `parse_specifier_for_upgrade` to explain the why.</sub>
